### PR TITLE
support record and list patterns in `prefer_final_locals`

### PR DESCRIPTION
**Pull request by pq**
Fixes: https://github.com/dart-lang/linter/issues/4120

/cc @bwilkerson

### CODE REVIEW

**Comment by pq** [comment]
Ah, I should probably handle object patterns too :/

**Comment by pq** [comment]
And Maps!

**Comment by pq** [comment]
OK. A bunch more cases handled.

@bwilkerson: feedback welcome.

**Review by srawlins — APPROVED** [review]
❤️

File: lib/src/rules/prefer_final_locals.dart [hunk 1]
```diff
@@ -76,6 +81,101 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+    }
+  }
+
+  bool isDeclaredFinal(AstNode node) {
+    var declaration = node.thisOrAncestorOfType<PatternVariableDeclaration>();
+    if (declaration == null) return false; // To be safe.
+    return declaration.keyword.isFinal;
+  }
+
+  bool isPotentiallyMutated(AstNode pattern, FunctionBody function) {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_final_locals.dart:129]: This is attempting to return `true` if any variable in the `pattern` is potentially mutated, but it doesn't appear to account for variables nested inside other patterns (such as a logical-and or list pattern).

File: lib/src/rules/prefer_final_locals.dart [hunk 2]
```diff
@@ -76,6 +81,101 @@ class _Visitor extends SimpleAstVisitor<void> {
...
 
+  void checkPatternFields(DartPattern node) {
+    var function = node.thisOrAncestorOfType<FunctionBody>();
+    if (function == null) return;
+
+    late NodeList<PatternField> fields;
+    if (node is RecordPattern) fields = node.fields;
+    if (node is ObjectPattern) fields = node.fields;
+
+    var parent = node.unParenthesized.parent;
```
Comment (bwilkerson) [inline lib/src/rules/prefer_final_locals.dart:92]: I don't think this is doing what you expect it to do. If the `node` is a `ParenthesizedPattern`, then `unParenthesized` will return the expression inside the parentheses, and `parent` will return the `ParenthesizedPattern`. I think you either want `node.parent` or a loop that will get you to the parent of any enclosing parenthesized patterns.

File: lib/src/rules/prefer_final_locals.dart [hunk 3]
```diff
@@ -100,3 +200,23 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+extension on Token? {
+  bool get isFinal {
+    var self = this;
+    if (self == null) return false;
+    return self.keyword == Keyword.FINAL;
+  }
+}
+
+extension on DartPattern {
+  DeclaredVariablePattern? get declaredVariablePattern {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_final_locals.dart:213]: I don't understand the semantics you're trying to implement here.

File: test/rules/prefer_final_locals_test.dart [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  test_destructured_listPattern_mutated_ok() async {
+    await assertNoDiagnostics(r'''
+f() {
+  var [a, b, ...rest] = [1, 2, 3, 4, 5, 6, 7];
+  print('${++a}$b$rest');
+}
+''');
+  }
+
+  test_destructured_listPattern_ok() async {
```
Comment (bwilkerson) [inline test/rules/prefer_final_locals_test.dart:32]: I'd drop the `_ok` (I think the `assertNoDiagnostics` says enough), especially here where we're expecting a lint.

File: test/rules/prefer_final_locals_test.dart [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  test_switch_objectPattern() async {
+    await assertDiagnostics(r'''
+class A {
+  int a;
+  A(this.a);
+}
+
+f() {
+  switch (A(1)) {
+    case A(a: >0 && var b): print('$b');
```
Comment (bwilkerson) [inline test/rules/prefer_final_locals_test.dart:180]: What happens when you reverse the operands of the `&&`?

File: lib/src/rules/prefer_final_locals.dart [hunk 6]
```diff
@@ -76,6 +81,101 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+      if (element == null || function.isPotentiallyMutatedInScope(element)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @override
+  void visitListPattern(ListPattern node) {
+    if (isDeclaredFinal(node)) return;
```
[the remark this answers is not in the document]
Comment (srawlins) [reply lib/src/rules/prefer_final_locals.dart:141]: Addressed in https://github.com/dart-lang/linter/pull/4225

File: lib/src/rules/prefer_final_locals.dart (hunk 2 above)
Comment (srawlins) [reply lib/src/rules/prefer_final_locals.dart:92]: Addressed in https://github.com/dart-lang/linter/pull/4225

File: lib/src/rules/prefer_final_locals.dart (hunk 1 above)
Comment (srawlins) [reply lib/src/rules/prefer_final_locals.dart:129]: Addressed in https://github.com/dart-lang/linter/pull/4225

File: lib/src/rules/prefer_final_locals.dart (hunk 3 above)
Comment (srawlins) [reply lib/src/rules/prefer_final_locals.dart:213]: Removed in https://github.com/dart-lang/linter/pull/4225

File: test/rules/prefer_final_locals_test.dart (hunk 4 above)
Comment (srawlins) [reply test/rules/prefer_final_locals_test.dart:32]: Done in https://github.com/dart-lang/linter/pull/4225

**Pull request merged**